src/montgomery.rs

```diff
@@ -91,6 +92,24 @@ impl PartialEq for MontgomeryPoint {
 
 impl Eq for MontgomeryPoint {}
 
+impl ValidityCheck for MontgomeryPoint {
+    /// Decode the \\(u\\)-coordinate field element and re-encode it
+    /// to its canonical form to check whether the original was valid.
+    ///
+    /// There are no other required checks for the Mongomery form of the curve,
+    /// as every element in \\( \mathbb{F}\_{q} \\) lies either on the curve or
+    /// its quadratic twist.  (cf. §5.2 of "Montgomery Curves and Their
+    /// Arithmetic" by [Costello and Smith][costello-smith].)
+    ///
+    /// [costello-smith]: https://eprint.iacr.org/2017/212.pdf
+    fn is_valid(&self) -> bool {
+        let maybe_u: FieldElement = FieldElement::from_bytes(&self.0);
+        let u: [u8; 32] = maybe_u.to_bytes();
+
+        u.ct_eq(&self.0).into()
+    }
+}
```

Comment:
§5.2 of "Montgomery Curves and Their Arithmetic" reads:

> both E_{(A,B)}(𝔽_q) and E_{(A,B')}(𝔽_q) have cryptographically strong (i.e. almost prime) group orders. This means that every element of 𝔽_q corresponds to a point on a cryptographically strong Montgomery curve, and implementers need not perform any point validation checks in this protocol.

My reading of that is that we do not need to check that a 32-byte string is a canonical representation of a field element, however we probably should check (which the decoding function explicitly does not do) in order to avoid inputs in [2^255 - 18, 2^255).

Reply:
That **is** the check that the encoding is canonical, which is why we don't need it.

Reply:
Yes, I realise that *is* the the check that the encoding is canonical. I'm arguing that we _should_ do the check to ensure the 32-bytes are a canonical element in 𝔽_q, otherwise the callee must check this if they care about potential "non-contributory" inputs. (My understanding of `TryFrom` is essentially "here's some random crap, make it absolutely safe if you can".)

Reply:
I mean, consumers of our library are doing this check on their own (cf. https://github.com/tendermint/kms/pull/279). I'm arguing that we should give them proper tools to Just Do The Thing And Not Have To Worry, since the whole point of this PR is to reduce copy-pasta.